src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -1397,23 +1424,25 @@ static int do_map(int argc, const char *argv[], Config *cfg)
 
   size = info.size;
 
-  r = load_module(cfg);
+  server = start_server(fd[1], image, cfg);
+
+retry:
+  r = load_module(cfg, remove_module);
   if (r < 0)
     goto close_fd;
 
-  server = start_server(fd[1], image, cfg);
-
-  use_netlink = cfg->try_netlink;
-  if (use_netlink) {
+  if (!cfg->use_ioctl) {
     r = try_netlink_setup(cfg, fd[0], size, flags);
     if (r < 0) {
       goto free_server;
     } else if (r == 1) {
-      use_netlink = false;
+      // Netlink interface is not supported and
+      // fallback to use ioctl interface.
+      cfg->use_ioctl = true;
+      remove_module = true;
```

Comment:
I am not very happy with the idea of reloading the module.

Actually, I think it was not very good idea to add the code to load the module initially -- I would not rely on this in production and prefer if the module with desirable params had been loaded before rbd-nbd was used, so we would avoid races like the one we try to fix here.

But if we already have the automatic module loading historically, let it be. And it may be useful in non-production environments. But now adding reload... It is not supposed to work reliably -- the module may be already in use. I expect even more races than we have now. And this is needed for the legacy case only, which we should not bother at all (I hope we just remove this code in the future).

Reply:
While I concur that I don't want to see the module unloaded, the fact that we load it definitely needs to stay. For example, as a storage admin deploying "ceph-csi" with the intention to use "rbd-nbd" instead of "krbd", I don't have immediate control over the kernel command-line or even potentially the ability to take a maintenance window to reboot w/ a new setting.

Reply:
Okay, will remove it.

While currently it will only remove it only when no any user is using it.

Reply:
Sure, I don't think we have to remove the possibility to load the module, as we already have it and users rely on it.

I was just telling that I would try to avid using this feature if possible. E.g. for your example with "ceph-csi", I would prefer a solution when the hosts with "ceph-csi" role assigned (i.e. where the nbd potentially can be used) just would always load the module on the host start.

Reply:
Maybe it is confusing with that, so let's remove it.